presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -328,23 +323,51 @@ private <T extends Node> T processSameType(T node)
 
     private QuerySpecification rewriteQuerySpecificationIfCompatible(QuerySpecification querySpecification, Table baseTable)
     {
-        QualifiedObjectName tableName = createQualifiedObjectName(session, baseTable, baseTable.getName());
-        List<ConnectorMaterializedViewDefinition> linkedMaterializedViews = baseTableToMaterializedViews.getOrDefault(tableName, ImmutableList.of());
+        Optional<String> errorMessage = MaterializedViewRewriteQueryShapeValidator.validate(querySpecification);
+
+        if (errorMessage.isPresent()) {
+            log.info(errorMessage.get());
+            return querySpecification;
+        }
+
+        List<QualifiedObjectName> referencedMaterializedViews = metadata.getReferencedMaterializedViews(
+                session,
+                createQualifiedObjectName(session, baseTable, baseTable.getName()));
 
-        for (ConnectorMaterializedViewDefinition materializedViewDefinition : linkedMaterializedViews) {
-            Table materializedViewTable = new Table(QualifiedName.of(materializedViewDefinition.getTable()));
-            Query materializedViewQuery = (Query) sqlParser.createStatement(materializedViewDefinition.getOriginalSql(), createParsingOptions(session));
+        // TODO: Select the most compatible and efficient materialized view for query rewrite optimization https://github.com/prestodb/presto/issues/16431
+        // TODO: Refactor query optimization code https://github.com/prestodb/presto/issues/16759
 
-            QuerySpecification rewrittenQuerySpecification = new QuerySpecificationRewriter(materializedViewTable, materializedViewQuery).rewrite(querySpecification);
+        for (QualifiedObjectName materializedViewName : referencedMaterializedViews) {
+            QuerySpecification rewrittenQuerySpecification = getRewrittenQuerySpecification(metadata, materializedViewName, querySpecification);
 
-            if (rewrittenQuerySpecification != querySpecification) {
+            if (rewrittenQuerySpecification == querySpecification) {
+                continue;
+            }
+            if (!isMaterializedViewDataConsistencyEnabled(session)) {
                 session.getRuntimeStats().addMetricValue(OPTIMIZED_WITH_MATERIALIZED_VIEW_SUBQUERY_COUNT, NONE, 1);
                 return rewrittenQuerySpecification;
             }
+            // TODO: We should be able to leverage this information in the StatementAnalyzer as well.
```

Comment:
@jainxrohit do we need this todo?

Reply:
Yes, we need this, right now we make call to getMaterializedViewStatus 2 times. We should be able to optimize it.